### PR TITLE
[Feature](bangc-ops): add fft op

### PR DESCRIPTION
**Pull request by SunsadYaofas**
Thanks for your contribution and we appreciate it a lot. :rocket::rocket:

## 1. Motivation

Please describe your motivation and the goal you want to achieve through this pull request.

## 2. Modification

Please briefly describe what modification is made in this pull request, and indicate where to make the modification.

Are new test cases added? If so, please post the corresponding generator-PR link here.

## 3. Test Report

If you want to know how to do operator testing, you can see [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).

#### 3.1.1 Accuracy Acceptance Standard

For static threshold standard details, see: [MLU-OPS Accuracy Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Accuracy-Acceptance-Standard.md).

- static threshold
  - diff1
    - [ ] float32 mlu diff1 <= 1e-5
    - [ ] float32 mlu diff1 <= 3e-3
    - [ ] float16 mlu diff1 <= 3e-3
  - diff2
    - [ ] float32 mlu diff2 <= 1e-5
    - [ ] float32 mlu diff2 <= 3e-3
    - [ ] float16 mlu diff2 <= 3e-3
  - diff3
    - [ ] mlu diff3 == 0
    - [ ] mlu diff3_1 == 0
    - [ ] mlu diff3_2 == 0
- dynamic threshold
  - [ ] diff1: mlu diff1 <= max(baseline diff1 * 10, static threshold)
  - [ ] diff2: mlu diff2 <= max(baseline diff2 * 10, static threshold)
  - [ ] diff3: mlu diff3 <= max(baseline diff3 * 10, static threshold)
    - float32, threshold = 1e-5
    - float16, threshold = 1e-3

#### 3.1.2 Operator Scheme checklist

- Supported hardware
  - [ ] MLU370
  - [ ] MLU590
- Job types
  - [ ] BLOCK
  - [ ] UNION1
  - [ ] UNION2
  - [ ] UNION4
  - [ ] The operator will dynamically select the most suitable task type, for example, UNION8

#### 3.2.1 Accuracy Test

If you have checked the following items, please tick the relevant box.

- [ ] Data type test (e.g. float32/int8)
- [ ] Multi-dimensional tensor test
- [ ] Layout test
- [ ] Different size/integer remainder end segment/alignment misalignment test
- [ ] Zero dimensional tensor test/zero element test
- [ ] stability test
- [ ] Multiple platform test
- [ ] Gen_case module test, see: [Gencase-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/Gencase-User-Guide-zh.md)
- [ ] Nan/INF tests 
- [ ] Bug fix tests
- [ ] For memory leak check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md)
- [ ] For code coverage check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md)
- [ ] For I/O calculation efficiency check details, see: [MLU-OPS-Performance-Acceptance-Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md)

#### 3.2.2 Parameter Check

Test Point-1: `When a new operator is submitted, the test points are given and the test results are stated`. Acceptance Standard: `Normal error`.
```bash
Please fill your test results(Error Message) in here, ...
```

Test Point-2: `Whether illegal parameters are passed`. Acceptance Standard: `Normal error`.
```bash
Test results...
```

### 3.3 Performance Test

See [MLU-OPS Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md) for details.

Platform：MLU370

```bash
# The test results should contain Op name, Shape, Data type,  
#   MLU Hardware Time(us), MLU Interface Time(us), MLU IO Efficiency, 
#   MLU Compute Efficiency, and Mlu Workspace Size(Bytes)
# 
# for example:
#
# ----------- case0 -----------
# case0
# [Op name                ]: abs
# [Shape                  ]: input.shape=[1024,1024,3,4], output.shape=[1024,1024,3,4]
# [Data type]             ]: float32
# [MLU Hardware Time      ]: 15728 (us)
# [MLU Interface Time     ]: 369.008 (us)
# [MLU IO Efficiency      ]: 0.23275
# [MLU Compute Efficiency ]: 0.5
# [Mlu Workspace Size     ]: -1 (Bytes)
# 
# ----------- case1 -----------
# ...
```

Platform：MLU590
```bash
# ----------- case0 -----------
# ----------- case1 -----------
# ...
```

### 3.4 Summary Analysis

Please give a brief overview here, if you want to note and summarize the content.

### CODE REVIEW

File: bangc-ops/kernels/fft/README.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
+# Introducation
+
+在这个目录中，我们将整合所有与FFT算子相关的代码，代码支持了FFT算子四种模式：rfft(r2c), irfft(c2r), fft(c2c), ifft(c2c)；同时，出于性能考虑，在具体实现中，不同规模会调用不同的算法，共有三种：DFT、FFT_cooley-tukey和FFT_stockham，具体的代码组织方式如下：
+
+## 代码目录以及说明
+
+1.目录的树状图如下：
+  ├── c2c_fft
```
Comment (PetrelYy) [inline bangc-ops/kernels/fft/README.md:8]: 乱码了，建议网页检查下markdown 格式

File: docs/bangc-docs/design_docs/fft/fft.md [hunk 2]
```diff
@@ -0,0 +1,736 @@
+# mluopFFT1d 算子开发设计方案
+
+
+* #### 文档基本信息
+
+| ----------- | ------------------------------------------------------------ |
```
Comment (PetrelYy) [inline docs/bangc-docs/design_docs/fft/fft.md:6]: markdown 格式问题统一修改
此外，还需在 
1. mlu_op.h 中添加算子接口+注释
2. docs 目录下 .rst 文件中添加算子介绍。文档部分 @suhaoming 同学review，详见开发流程

File: bangc-ops/kernels/utils/common.h [hunk 3]
```diff
@@ -686,4 +686,67 @@ __mlu_func__ void __mluop_get_stage_indices_tfuse(int *dst_nram,
 #endif
 }
 
+template <typename T>
+__mlu_func__ void __mluops_getMToNIndices(const uint32_t num_integers,
+                                        T first_index,
```
Comment (PetrelYy) [inline bangc-ops/kernels/utils/common.h:691]: 未对齐，可使用mlu-ops/tools/format 工具刷格式

File: bangc-ops/kernels/utils/common.h [hunk 4]
```diff
@@ -686,4 +686,67 @@ __mlu_func__ void __mluop_get_stage_indices_tfuse(int *dst_nram,
 #endif
 }
 
+template <typename T>
+__mlu_func__ void __mluops_getMToNIndices(const uint32_t num_integers,
```
Comment (PetrelYy) [inline bangc-ops/kernels/utils/common.h:690]: 接口标识是 __mluop_* ，函数名使用下划线格式，不建议与驼峰混用

File: bangc-ops/kernels/fft/c2c_fft/c2c_fft.h [hunk 5]
```diff
@@ -0,0 +1,32 @@
+/*************************************************************************
+ * Copyright (C) [2019-2022] by Cambricon, Inc.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
```
Comment (PetrelYy) [inline bangc-ops/kernels/fft/c2c_fft/c2c_fft.h:7]: 统一修改文件 license

File: bangc-ops/kernels/fft/c2c_fft/c2c_fft_host.mlu [hunk 6]
```diff
@@ -0,0 +1,1152 @@
...
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *************************************************************************/
+
+#include "kernels/fft/c2c_fft/c2c_fft.h"
+#include <algorithm>
+#include <string>
+
+#define DIRECTION 2  // FORWARD and BACKWARD
+
+mluOpStatus_t selectFFT1dStrategy(mluOpHandle_t handle, mluOpFFTPlan_t fft_plan) {
```
Comment (PetrelYy) [inline bangc-ops/kernels/fft/c2c_fft/c2c_fft_host.mlu:19]: 只在本文件中使用的函数建议使用 static 修饰

File: bangc-ops/kernels/fft/c2c_fft/c2c_fft_host.mlu [hunk 7]
```diff
@@ -0,0 +1,1152 @@
...
+  const unsigned int cluster_number = mluOp::runtime::getClusterLimitCapability(handle);
+  const unsigned int core_dim = handle->core_num_per_cluster;
+  cnrtDim3_t k_dim = {core_dim, cluster_number, 1};
+  cnrtFunctionType_t k_type = CNRT_FUNC_TYPE_BLOCK;
+
+  switch (fft_plan->fft_strategy) {
+    case CNFFT_FUNC_MATMUL: {
+      // Matmul Matrix : 2 * 2 * [n, n] (forward and backward)
+      int dft_mat_num = dft_mat_times * n * n;
+      KERNEL_CHECK((generateC2CFFTDFTMatrix<<<k_dim, k_type, handle->queue>>>(
```
Comment (PetrelYy) [inline bangc-ops/kernels/fft/c2c_fft/c2c_fft_host.mlu:451]: 语法糖部分放在 device 代码文件里，host 代码以 .cpp 承载。 可参考其他算子

File: bangc-ops/kernels/fft/fft_optm_device/fft_cooley-tukey_ux_device.mlu [hunk 8]
```diff
@@ -0,0 +1,833 @@
...
+  float *inc_addr    = w_tmp2;
+
+  // deal with each L segment.
+  for (int i = 0; i < n / L_align; i++) {
+    float *tmp_cos_addr    = cos_addr    + L_align * i;
+    float *tmp_sin_addr    = sin_addr    + L_align * i;
+    float *tmp_offset_addr = offset_addr + L_align * i;
+    float *tmp_inc_addr    = inc_addr    + L_align * i;
+    float start_value      = L * i;
+#if __BANG_ARCH__ >= 322
```
Comment (PetrelYy) [inline bangc-ops/kernels/fft/fft_optm_device/fft_cooley-tukey_ux_device.mlu:53]: 建议改为372，ops 未支持端侧

File: bangc-ops/kernels/fft/fft_optm_device/fft_cooley-tukey_ux_device.mlu [hunk 9]
```diff
@@ -0,0 +1,833 @@
...
+    __mluOp_getMToNIndices(L_align, start_value, tmp_inc_addr, nullptr, tmp_inc_addr, nullptr);
+#else
+    float *int32_temp_addr = tmp_cos_addr;
+    float *float2int_dst_ad   = tmp_sin_addr;
```
Comment (PetrelYy) [inline bangc-ops/kernels/fft/fft_optm_device/fft_cooley-tukey_ux_device.mlu:57]: 格式

File: docs/bangc-docs/design_docs/fft/fft.md [hunk 10]
```diff
@@ -0,0 +1,736 @@
...
+
+```c++
+/*! The descriptor of FFT (Fast Fourier Transform) operator that holds FFT information including
+ * the tensor descriptor of input tensor and output tensor, the rank of FFT, the FFT size on each
+ * dimension, the size of reserved space and the size of workspace.
+ *
+ * You need to call the ::mluopCreateFFTPlan function to create a descriptor for the FFT operator, and call
+ * the ::mluopMakeFFTPlanMany function to set the information of the FFT operator to the descriptor.
+ * Then, you need to allocate the reserved space and set the space to the fft descriptor by ::mluopSetReserveArea.
+ * Also, you need to destroy the CNNL context at the end with the ::mluopDestroyFFTPlan.
```
Comment (PetrelYy) [inline docs/bangc-docs/design_docs/fft/fft.md:282]:
```suggestion
 * Also, you need to destroy the  MLU-OPS context at the end with the ::mluopDestroyFFTPlan.
```

File: bangc-ops/kernels/fft/README.md (hunk 1 above)
Comment (SunsadYaofas) [reply bangc-ops/kernels/fft/README.md]: 已经修改

File: docs/bangc-docs/design_docs/fft/fft.md (hunk 2 above)
Comment (SunsadYaofas) [reply docs/bangc-docs/design_docs/fft/fft.md:6]: 1、补充gtest
2、mlu_op.h 已经中添加算子接口+注释
3、

File: bangc-ops/kernels/utils/common.h (hunk 3 above)
Comment (SunsadYaofas) [reply bangc-ops/kernels/utils/common.h:691]: 已修改

File: bangc-ops/kernels/utils/common.h (hunk 4 above)
Comment (SunsadYaofas) [reply bangc-ops/kernels/utils/common.h]: 已修改

File: bangc-ops/kernels/fft/fft_optm_device/fft_cooley-tukey_ux_device.mlu (hunk 8 above)
Comment (SunsadYaofas) [reply bangc-ops/kernels/fft/fft_optm_device/fft_cooley-tukey_ux_device.mlu:53]: 已修改

File: bangc-ops/kernels/utils/common.h (hunk 3 above)
Comment (SunsadYaofas) [reply bangc-ops/kernels/utils/common.h:691]: mlu-ops/tools/format  报please install clang-format 怎么办，哪个卡可以支持？

File: bangc-ops/kernels/fft/c2c_fft/c2c_fft_host.mlu (hunk 7 above)
Comment (SunsadYaofas) [reply bangc-ops/kernels/fft/c2c_fft/c2c_fft_host.mlu:451]: 学习一下

File: bangc-ops/kernels/fft/fft_optm_device/fft_cooley-tukey_ux_device.mlu (hunk 9 above)
Comment (SunsadYaofas) [reply bangc-ops/kernels/fft/fft_optm_device/fft_cooley-tukey_ux_device.mlu:57]: 已经刷新

File: docs/bangc-docs/design_docs/fft/fft.md (hunk 10 above)
Comment (SunsadYaofas) [reply docs/bangc-docs/design_docs/fft/fft.md]: 已经完成

File: bangc-ops/kernels/fft/README.md (hunk 1 above)
Comment (PetrelYy) [reply bangc-ops/kernels/fft/README.md]: 还是存在问题，├── c2c_fft 错位了，这里可以使用代码段

File: bangc-ops/mlu_op.h [hunk 11]
```diff
@@ -16613,6 +16613,270 @@ mluOpQuantizeBatchMatMulBCast(mluOpHandle_t handle,
...
+    @par Return
+    - ::CNNL_STATUS_SUCCESS, ::CNNL_STATUS_ALLOC_FAILED
+    *
+    @par API Dependency
+    - After calling this function, you can call the ::mluOpMakeFFTPlanMany function to initialize and set the
+    information to the created descriptor.
+    - You need to call the ::mluOpDestroyFFTPlan to destroy the descriptor.
+    Otherwise, the memory leak may occur.
+    *
+    @note
```
Comment (PetrelYy) [inline bangc-ops/mlu_op.h:16646]: 接口注释格式请参考 https://wiki.cambricon.com/pages/viewpage.action?pageId=255658083

File: bangc-ops/kernels/utils/common.h (hunk 3 above)
Comment (PetrelYy) [reply bangc-ops/kernels/utils/common.h]: .25 可以直接使用

File: docs/bangc-docs/design_docs/fft/fft.md [hunk 12]
```diff
@@ -0,0 +1,732 @@
...
+
+#### 3.1.1 Iterative FFT：Stockham FFT Algorithm
+
+用来处理2的整数次幂的部分，Stockham FFT属于迭代式Cooley-tukey公式的变种，可以由Cooley Tukey FFT变换得到。长度为n的Cooley-tukey计算公式如下：
+```math
+\text{DFT}_n=(\text{DFT}_k \otimes I_m) T_m^n(I_k \otimes \text{DFT}_m) L_k^n,n=km,\tag{1}
+```
+其中$`\otimes`$表示为克罗内克乘积，$`(\text{DFT}_k \otimes I_m)`$为向量并行矩阵；$`T_m^n`$为旋转因子，为对角矩阵；$`(I_k \otimes \text{DFT}_m)`$为块内并行矩阵；$`L_k^n`$形式为$`L_m^{mn}:={in+j \rarr jm+i, 0\le i<m, 0 \le j < n}`$ 的置换矩阵。下面展示了 $`k=4,m=2`$下的矩阵：
+```math
+I_4 \otimes \text{DFT}_2=\left[
```
Comment (PetrelYy) [inline docs/bangc-docs/design_docs/fft/fft.md:427]: 公式挂了

File: bangc-ops/kernels/fft/README.md (hunk 1 above)
Comment (SunsadYaofas) [reply bangc-ops/kernels/fft/README.md]: 已经修改

File: docs/bangc-docs/design_docs/fft/fft.md (hunk 12 above)
Comment (SunsadYaofas) [reply docs/bangc-docs/design_docs/fft/fft.md:427]: 已修改

File: bangc-ops/mlu_op.h (hunk 11 above)
Comment (SunsadYaofas) [reply bangc-ops/mlu_op.h]: 已修改 && 找suhaoming review中

File: bangc-ops/mlu_op.h [hunk 13]
```diff
@@ -16613,6 +16613,271 @@ mluOpQuantizeBatchMatMulBCast(mluOpHandle_t handle,
...
+    *
+    @par Note
+    - This function only supports 1D FFT currently. 2D FFT and 3D FFT
+    will be supported in the future.
+    - When the data type of input is float or complex_float, the 1D FFT length should be equal to:
+    length = \f$base * 2^ {m}\f$ , and the base should be less than or equal to 4096.
+    * - When the data type of input is half or complex_half, the 1D FFT length should be equal to:
+    length = \f$2^{m}
+
+    \f$.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:16655]: 公式这么写可能会报错呀

File: bangc-ops/mlu_op.h [hunk 14]
```diff
@@ -16613,6 +16613,271 @@ mluOpQuantizeBatchMatMulBCast(mluOpHandle_t handle,
...
+    @par Example
+    - None.
+    */
+mluOpStatus_t
+mluOpCreateFFTPlan(mluOpFFTPlan_t *fft_plan);
+
+// Group:FFT
+/*!
+
+    @brief Initializes the FFT descriptor pointed by \p fft_plan that is previously created
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:16669]:
```suggestion
    @brief Initializes the FFT descriptor pointed by \p fft_plan, which is created
```

File: bangc-ops/mlu_op.h [hunk 15]
```diff
@@ -16613,6 +16613,271 @@ mluOpQuantizeBatchMatMulBCast(mluOpHandle_t handle,
...
+    with the ::mluOpCreateFFTPlan function, and sets the information about the
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:16670]:
```suggestion
    by ::mluOpCreateFFTPlan, and sets the information about the
```

File: bangc-ops/mlu_op.h [hunk 16]
```diff
@@ -16613,6 +16613,271 @@ mluOpQuantizeBatchMatMulBCast(mluOpHandle_t handle,
...
+    of input tensor descriptor.
+    *
+    @par Note
+    - The advanced data layout parameters, (i/o)nembed, (i/o)istride, (i/o)idist, are set through
+    ::mluOpSetTensorDescriptorEx. If stride information is not needed, you can set the simple data layout
+    through ::mluOpSetTensorDescriptor.
+    - The dimension size of input / output should be equal to \p rank or \p rank + 1. In the former case,
+    the batch size is considered as 1. Otherwise, the outermost dimension is the batch size.
+    - For real-to-complex FFTs, the innermost dimension of FFT length and output arrays are not the same.
+    For a x-length 1D real-to-complex FFT, the output is x/2 + 1 complex numbers (the non-redundant outputs).
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:16742]:
```suggestion
    For an x-length 1D real-to-complex FFT, the output is x/2 + 1 complex numbers (the non-redundant outputs).
```

File: bangc-ops/mlu_op.h [hunk 17]
```diff
@@ -16613,6 +16613,271 @@ mluOpQuantizeBatchMatMulBCast(mluOpHandle_t handle,
...
+    For a N-D real-to-complex FFT with n=[z, y, x], the output shape will be [z, y, x/2+1].
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:16743]:
```suggestion
    For an N-D real-to-complex FFT with n=[z, y, x], the output shape will be [z, y, x/2+1].
```

File: bangc-ops/mlu_op.h [hunk 18]
```diff
@@ -16613,6 +16613,271 @@ mluOpQuantizeBatchMatMulBCast(mluOpHandle_t handle,
...
+    - For complex-to-real FFTs, the input tensor only holds the non-redundant part of the Fourier coefficients.
+    And the output tensor stores the real output values.
+    - When n[0] is greater than 4096, the data type of input only supports float or complex_float.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:16746]: float or complex_float 确认一下这俩如果是代码就用 \p 的格式，如果是描述就用全称比如 floating point data

File: bangc-ops/mlu_op.h [hunk 19]
```diff
@@ -16613,6 +16613,271 @@ mluOpQuantizeBatchMatMulBCast(mluOpHandle_t handle,
...
+                     const mluOpTensorDescriptor_t output_desc,
+                     const int rank,
+                     const int n[],
+                     size_t *reservespace_size,
+                     size_t *workspace_size);
+
+// Group:FFT
+/*!
+
+    @brief Bond the \p reservespace to the \p fft_plan. The size of reserved space can be derived
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:16768]:
```suggestion
    @brief Binds the \p reservespace to the \p fft_plan. The size of the reserved space can be derived
```

File: bangc-ops/mlu_op.h [hunk 20]
```diff
@@ -16613,6 +16613,271 @@ mluOpQuantizeBatchMatMulBCast(mluOpHandle_t handle,
...
+    Input/Output. The descriptor of FFT. For detailed information, see ::mluOpFFTPlan_t.
+    @param[in] reservespace
+    Input. Pointer to the MLU memory that is used as an extra memory space for saving
+    intermediate results of FFT operation.
+    *
+    @par API Dependency
+    - Before calling this function, you need to call the ::mluOpCreateFFTPlan function
+    to create an FFT descriptor firstly, call the ::mluOpMakeFFTPlanMany function to set the
+    FFT descriptor and get the size of reserved space, and then call the
+    cnrtMalloc function to create MLU memory according to the rservespace_size given.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:16784]: 再确认一下，感觉这个参数名写错了

File: bangc-ops/mlu_op.h [hunk 21]
```diff
@@ -16613,6 +16613,271 @@ mluOpQuantizeBatchMatMulBCast(mluOpHandle_t handle,
...
+    @par Return
+    - ::MLUOP_STATUS_SUCCESS, ::MLUOP_STATUS_BAD_PARAM, ::MLUOP_STATUS_INTERNAL_ERROR
+    */
+mluOpStatus_t
+mluOpSetFFTReserveArea(mluOpHandle_t handle, mluOpFFTPlan_t fft_plan, void *reservespace);
+
+// Group:FFT
+/*!
+
+    @brief Executes any FFT. In case of complex-to-real and real-to-complex
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:16795]: 是这个意思么？Executes an arbitrary FFT.

File: bangc-ops/mlu_op.h [hunk 22]
```diff
@@ -16613,6 +16613,271 @@ mluOpQuantizeBatchMatMulBCast(mluOpHandle_t handle,
...
+    transforms \p direction parameter is ignored. This function stores the Fourier coefficients
+    in the output array. If the address of input and output are the same, an in-place FFT
+    is adopted.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:16798]: 是不是performed更合适一点儿？

File: bangc-ops/mlu_op.h [hunk 23]
```diff
@@ -16613,6 +16613,271 @@ mluOpQuantizeBatchMatMulBCast(mluOpHandle_t handle,
...
+    *
+    @param[in] handle
+    Input. Handle to a Cambricon MLUOP context that is used to manage MLU devices and queues
+    in the FFT execution. For detailed information, see ::mluOpHandle_t.
+    @param[in] fft_plan
+    Input. The plan for FFT execution. For detailed information, see ::mluOpFFTPlan_t.
+    @param[in] input
+    Input. Pointer to the MLU memory that stores the input tensor.
+    @param[in] scale_factor
+    Input. A float-point scalar used to multiply the FFT output.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:16808]: floating-point ？

File: bangc-ops/mlu_op.h [hunk 24]
```diff
@@ -16613,6 +16613,271 @@ mluOpQuantizeBatchMatMulBCast(mluOpHandle_t handle,
...
+    @param[in, out] workspace
+    Input/Output. Pointer to the MLU memory that is used as an extra workspace for the
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:16810]:
```suggestion
    Input/Output. Pointer to the MLU memory that is used as an additional workspace for 
```

File: bangc-ops/mlu_op.h [hunk 25]
```diff
@@ -16613,6 +16613,271 @@ mluOpQuantizeBatchMatMulBCast(mluOpHandle_t handle,
...
+    ::mluOpExecFFT.
+    @param[out] output
+    Output. Pointer to the MLU memory that stores the output tensor.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:16813]:
```suggestion
    Pointer to the MLU memory storing the output tensor.
```

File: bangc-ops/mlu_op.h [hunk 26]
```diff
@@ -16613,6 +16613,271 @@ mluOpQuantizeBatchMatMulBCast(mluOpHandle_t handle,
...
+    @param[in] direction
+    Input. The transform direction: 0 means FFT forward and 1 means FFT inverse.
+    Direction is ignored for real-to-complex and complex-to-real transforms.
+    *
+    @par Note
+    - For in-place 1D real-to-complex FFTs, the input is a batch of n real numbers, and the
+    output is n/2 + 1 non-redundant complex numbers. This requires a padding of input array.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:16820]:
```suggestion
    output is n/2 + 1 non-redundant complex numbers. This requires a padding of the input array.
```

File: bangc-ops/mlu_op.h [hunk 27]
```diff
@@ -16613,6 +16613,271 @@ mluOpQuantizeBatchMatMulBCast(mluOpHandle_t handle,
...
+    - For in-place N-D real-to-complex FFTs, extra padding of the real-data array on the innermost
+    dimension is necessary to accommodate the size of the complex-data output.
+    - When \p input contains NaN or infinity and the input onchip data type of FFT is not quantized
+    data type, the output is computed through the fft formula with computation rules of NaN or
+    infinity based on IEEE754.
+    - When \p input contains NaN or infinity and the input onchip data type of FFT is quantized
+    data type, i.e. int31 or int16, the output will be unpredictable.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:16827]: i.e. 还是 e.g.呀？

File: bangc-ops/mlu_op.h [hunk 28]
```diff
@@ -16613,6 +16613,271 @@ mluOpQuantizeBatchMatMulBCast(mluOpHandle_t handle,
...
+    - The range of \p input is recommended to be in range of [-10, 10] with uniform
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:16828]:
```suggestion
    - It is recommended that  \p input is in range of [-10, 10] with uniform
```

File: bangc-ops/mlu_op.h [hunk 29]
```diff
@@ -16613,6 +16613,271 @@ mluOpQuantizeBatchMatMulBCast(mluOpHandle_t handle,
...
+    distribution for higher precision.
+    - The range of \p scale_factor is recommended to be in range of [-1, 1] to avoid exceeding
+    the data representation range.
+    - Half data type of \p input is not recommended due to low precision. The first element of the
+    FFT result is the sum of all input elements, and it is likely to overflow.
+    - This operation is not supported on the 1V platforms.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:16834]: 1V platform？ 是不是改成架构号

File: bangc-ops/mlu_op.h [hunk 30]
```diff
@@ -16613,6 +16613,271 @@ mluOpQuantizeBatchMatMulBCast(mluOpHandle_t handle,
...
+             void *output,
+             int direction);
+
+// Group:FFT
+/*!
+
+    @brief Destroys an FFT plan \p fft_plan that is created with the
+    ::mluOpCreateFFTPlan function.
+    *
+    The fft plan is defined in ::mluOpFFTPlan_t and holds the information about the
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:16860]: 注意大小写哦

**Review by niyuming — APPROVED** [review]

**Pull request closed without merging**